storage/src/main/java/org/apache/kafka/storage/internals/log/UnifiedLog.java

```diff
@@ -16,44 +16,2260 @@
+                        // lookup the position of batch to avoid extra I/O
+                        OffsetPosition position = latestTimestampSegment.offsetIndex().lookup(maxTimestampSoFar.offset);
+                        Optional<FileRecords.TimestampAndOffset> timestampAndOffsetOpt =
+                                StreamSupport.stream(latestTimestampSegment.log().batchesFrom(position.position).spliterator(), false)
+                                    .filter(b -> b.maxTimestamp() == maxTimestampSoFar.timestamp)
+                                    .findFirst()
+                                    .flatMap(batch -> batch.offsetOfMaxTimestamp()
+                                        .map(offset -> new FileRecords.TimestampAndOffset(
+                                            batch.maxTimestamp(),
+                                            offset,
+                                            Optional.of(batch.partitionLeaderEpoch()).filter(epoch -> epoch >= 0))));
+                        return new OffsetResultHolder(timestampAndOffsetOpt);
+                    } else {
+                        // We need to search the first segment whose largest timestamp is >= the target timestamp if there is one.
+                        if (remoteLogEnabled() && !isEmpty()) {
+                            if (remoteOffsetReader.isEmpty()) {
+                                throw new KafkaException("RemoteLogManager is empty even though the remote log storage is enabled.");
+                            }
+
+                            AsyncOffsetReadFutureHolder<OffsetResultHolder.FileRecordsOrError> asyncOffsetReadFutureHolder =
+                                    remoteOffsetReader.get().asyncOffsetRead(topicPartition(), targetTimestamp,
+                                    logStartOffset, leaderEpochCache, () -> searchOffsetInLocalLog(targetTimestamp, localLogStartOffset()));
+
+                            return new OffsetResultHolder(Optional.empty(), Optional.of(asyncOffsetReadFutureHolder));
+                        } else {
+                            return new OffsetResultHolder(searchOffsetInLocalLog(targetTimestamp, logStartOffset));
+                        }
+                    }
+                });
+    }
+
+    /**
+     * Checks if the log is empty.
+     * @return Returns True when the log is empty. Otherwise, false.
+     */
+    public boolean isEmpty() {
+        return logStartOffset == logEndOffset();
+    }
+
+    private Optional<FileRecords.TimestampAndOffset> searchOffsetInLocalLog(long targetTimestamp, long startOffset) {
+        // Cache to avoid race conditions.
+        List<LogSegment> segmentsCopy = List.copyOf(logSegments());
+        Optional<LogSegment> targetSeg = segmentsCopy.stream().filter(s -> {
+            try {
+                return s.largestTimestamp() >= targetTimestamp;
+            } catch (IOException e) {
+                return false;
+            }
+        }).findFirst();
```

Comment:
TODO: This always iterates over the full collection, whereas the Scala find() method returns as soon as an element matching the predicate is found.